### ISSUE TITLE
CommunicationObject.OnOpen and OnBeginOpen is never called

### ISSUE THREAD

**Issue by fujiy** [dotnet/wcf#1544]
I don't know if it is a bug or intended, but the behavior is a break changing from .NET 4.5.

UPDATE: I checked the history. There is a commit to fix it: https://github.com/dotnet/wcf/blob/ab17582afcd91cec8cb286460793d4258015a2b3/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs#L776

But when you see the lastest version, it is not there:
https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

Thanks

**Comment by roncain** [comment]
Thanks @fujiy -- could I ask for some more details of your code that experiences this?  For example, I presume you have a custom channel or something like that.  As you saw, we are moving to a Task-based model internally that uses OnOpenAsync.  The issue I think we're seeing here is probably a case where we don't correctly handle a custom CommunicationObject and are trying to maintain backwards compatibility with non-product code. I'd like to create a repro to see exactly how OnOpen and OnBeginOpen are implemented in that code.  Is that something you can provide?

**Comment by fujiy** [comment]
It is a customer's code. But this commit, do what we need: https://github.com/dotnet/wcf/commit/ab17582afcd91cec8cb286460793d4258015a2b3. OnOpen is called, so my custom Channel can initialize the connection. But OnBeginOpen isn't called yet.

So in this commit: https://github.com/dotnet/wcf/commit/f64114d28778aa3d84d7b35f25fd069ffdb68ba2#diff-e22ad1a28d1d9894587c6d21bb86cd06 we have the inverse, OnBeginOpen is called, and OnOpen isn't anymore.

Now, using the current version, I could made it work overring OnOpening, but doing it I always execute the sync version of my Connect().

Looking the 4.6 code, we had both been called (maybe because we had distinct 2 implementations, one for sync and another for async and now we have just one, Task based and the sync versions call the same method and then Wait() it):
OnOpen:
http://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Channels/CommunicationObject.cs,591

OnBeginOpen:
http://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Channels/CommunicationObject.cs,226
http://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Channels/CommunicationObject.cs,1104

**Comment by roncain** [comment]
Thanks @fujiy .  I think I see the issue, as you showed in the commits you quoted, and I am working on a fix.  There is some subtlety here, so simply going back to the prior commit is not sufficient for other scenarios.  If I create a PR that has the fix, are you in a position to try it (meaning enlist in the repo, build the product, and use the binaries it produces)?  Or are you required to wait for packages to be published on NuGet or MyGet?

**Comment by fujiy** [comment]
I'll try, I just don't know if I'll make it in time for your release. Because I work in different customers every week.
But I would like to contribute here.

**Comment by roncain** [comment]
Thanks @fujiy -- I am almost ready to test a PR for this.  From my investigations, the issue was only that OnOpen was not called when you used the synchronous Open().  But OnBeginOpen was called properly when you used the asynchronous BeginOpen.  Does that agree with what you found?  The issue title seems to indicate both OnOpen and OnBeginOpen are broken.

**Comment by roncain** [comment]
@fujiy -- PR #1547 has an initial draft of a fix, but we are considering other alternatives.  You might be able to help us choose the right approach if you can answer these questions:
- Where (in your code) is the OnOpen() that is not being called? Is it a custom channel you implemented, of have you found some place in the WCF product code where this failure occurs?
- Does this implementation also implement OnBeginOpen()?
- How is this CommunicationObject being opened?  For example, are you relying on the explicit Open() that occurs when you try to use it?  Are you calling Open() explicitly?  Are you calling BeginOpen()?

My early observation is that any CommunicationObject implementing OnBeginOpen() and OnBeginClose() should have worked properly, even without my bug fix.  It was only the synchronous OnOpen() and OnClose() that were broken. If this is not true for your scenario, then we need to understand your exact usage better to consider a different fix.  Take a look at the unit tests in this PR for a sense of the kinds of validation we are doing -- it might help suggest other validation we could do that matches your scenario better.

And thanks!  We appreciate your taking the time to raise this issue and help us find the right solution.

**Comment by fujiy** [comment]
@roncain 

It is a CustomChannel.
We have both implemented
I'll send you a simple implementation I created, so we can talk offline.

**Comment by roncain** [comment]
Thanks @fujiy .  We've checked in a fix that solved the one scenario where I was able to reproduce the issue.  I'll like to be able to verify against your scenario too.  One way to share code would be for you to create a gist [like this](https://gist.github.com/). There are also a couple options available to you to try out the fix against your real code:
1) Enlist in this repo, build the product, and use the binaries it produces.  I can tell you which ones if you want to try that.
2) Grab last night's packages, such as [this one](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.ServiceModel.Primitives)

Obviously, the binaries and packages are work in progress, so they are not suitable for production until we release stable versions to NuGet. It's just one alternative if you want to see the code under development.
